mmdet/models/dense_heads/decoupled_solo_head.py

```diff
@@ -0,0 +1,569 @@
+                                              size=seg_num_grid, mode='bilinear')
+                cate_feat = cate_layer(cate_feat)
+
+            cate_pred = self.dsolo_cate(cate_feat)
+            if eval:
+                ins_pred_x = F.interpolate(ins_pred_x.sigmoid(),
+                                           size=upsampled_size, mode='bilinear')
+                ins_pred_y = F.interpolate(ins_pred_y.sigmoid(),
+                                           size=upsampled_size, mode='bilinear')
+                cate_pred = points_nms(cate_pred.sigmoid(),
+                                       kernel=2).permute(0, 2, 3, 1)
+            ins_pred_x_maps.append(ins_pred_x)
+            ins_pred_y_maps.append(ins_pred_y)
+            cate_pred_maps.append(cate_pred)
+        return ins_pred_x_maps, ins_pred_y_maps, cate_pred_maps
+
+    def forward_train(self,
+                      x,
+                      img_metas,
+                      gt_bboxes,
+                      gt_labels=None,
+                      gt_bboxes_ignore=None,
+                      gt_mask=None,
+                      proposal_cfg=None,
+                      **kwargs):
+        outs = self(x)
+        if gt_labels and gt_mask is None:
+            loss_inputs = outs + (gt_bboxes, img_metas)
+        elif gt_labels and gt_mask is not None:
+            loss_inputs = outs + (gt_bboxes, gt_labels, img_metas, gt_mask)
+        elif gt_labels is not None and gt_mask is None:
+            loss_inputs = outs + (gt_bboxes, gt_labels, img_metas)
+        losses = self.loss(*loss_inputs, gt_bboxes_ignore=gt_bboxes_ignore)
+        if proposal_cfg is None:
+            return losses
+        else:
+            proposal_list = self.get_bboxes(*outs, img_metas, cfg=proposal_cfg)
+            return losses, proposal_list
+
+    def loss(self,
+             ins_preds_x,
+             ins_preds_y,
+             cate_preds,
+             gt_bbox_list,
+             gt_label_list,
+             img_metas,
+             gt_mask_list,
+             cfg=None,
+             gt_bboxes_ignore=None):
+        featmap_sizes = [featmap.size()[-2:] for featmap in
+                         ins_preds_x]
+
+        ins_label_list, cate_label_list, \
+        ins_ind_label_list, ins_ind_label_list_xy = multi_apply(
+            self.solo_target_single,
+            gt_bbox_list,
+            gt_label_list,
+            gt_mask_list,
+            featmap_sizes=featmap_sizes)
+
+        # ins
+        ins_labels = [torch.cat([ins_labels_level_img[
+                                     ins_ind_labels_level_img, ...]
+                                 for ins_labels_level_img, ins_ind_labels_level_img in
+                                 zip(ins_labels_level, ins_ind_labels_level)], 0)
+                      for ins_labels_level, ins_ind_labels_level in
+                      zip(zip(*ins_label_list), zip(*ins_ind_label_list))]
+
+        ins_preds_x_final = [torch.cat([ins_preds_level_img_x[
```

Comment:
should make lines 244-250 more clear, decouple the nested for loops, and use the conventional style like
```python
for xxx:
   for xxx
```